src/mapped-types.ts

```diff
@@ -412,12 +412,18 @@ export type PromiseType<T extends Promise<any>> = T extends Promise<infer U>
  *   };
  *   type ReadonlyNestedProps = DeepReadonly<NestedProps>;
  */
-export type DeepReadonly<T> = T extends (...args: any[]) => any
+export type DeepReadonly<T> = T extends
+  | ((...args: any[]) => any)
+  | string
+  | number
+  | boolean
+  | null
+  | undefined
   ? T
-  : T extends any[]
-  ? _DeepReadonlyArray<T[number]>
-  : T extends object
-  ? _DeepReadonlyObject<T>
+  : T extends _DeepReadonlyArray<infer U>
+  ? _DeepReadonlyArray<U>
+  : T extends _DeepReadonlyObject<infer V>
+  ? _DeepReadonlyObject<V>
```

Comment:
Can't we flip that so it will first match _DeepReadonlyArray & _DeepReadonlyObject? Then the rest seems to be always mapped to T anyway.
It would be much cleaner.

Reply:
If I understand what you're asking correctly, then no, `string` and `boolean` actually match to `_DeepReadonlyObject<infer V>`. It extracts out things like `toString()` and whatnot.

Reply:
Ok got it they need to short circuit before them.